clash-prelude/src/Clash/Sized/Fixed.hs

```diff
@@ -904,6 +905,182 @@ fLitR a = Fixed (fromInteger sat)
+
+  toEnum i =
+    let sh   = natToNum @frac
+        res  = toInteger i `shiftL` sh
+        rMax = toInteger (maxBound :: rep (int + frac))
+        rMin = toInteger (minBound :: rep (int + frac))
+    in if res > rMax || res < rMin then
+         error $  "Enum.toEnum{"
+               ++ show (typeRep $ Proxy @(Fixed rep int frac)) ++ "}: tag ("
+               ++ show i ++ ") is outside of bounds "
+               ++ show ( minBound :: Fixed rep int frac
+                       , maxBound :: Fixed rep int frac)
+       else
+         Fixed (fromInteger res)
+
+  fromEnum f@(Fixed fRep) =
+    let nF     = natToNum @frac
+        frMask = fromInteger $ (1 `shiftL` nF) - 1
+        offset = if f < 0 && fRep .&. frMask /= 0 then 1 else 0
+        -- res amounts to "truncate f", but without needing all the constraints
+        -- for RealFrac.
+        res    = toInteger $ (fRep `shiftR` nF) + offset
+        rMax   = toInteger (maxBound :: Int)
+        rMin   = toInteger (minBound :: Int)
+    in if res > rMax || res < rMin then
+         error $  "Enum.fromEnum{" ++ show (typeOf f) ++ "}: value ("
+               ++ show f ++ ") is outside of Int's bounds "
+               ++ show (rMin, rMax)
+       else
+         fromInteger res
+
+  enumFrom x1 = enumFromTo x1 maxBound
+  enumFromThen (Fixed x1Rep) (Fixed x2Rep) =
+    map Fixed $ enumFromThen x1Rep x2Rep
+
+  enumFromTo x1@(Fixed x1Rep) y@(Fixed yRep)
+    | yPlusHalf < x1 = []
+    | closeToMax     = [x1]
+    | otherwise      =  map Fixed $ enumFromThenTo
+                                      x1Rep
+                                      (unFixed $ satSucc SatWrap x1)
+                                      (unFixed $ yPlusHalf)
+   where
+    closeToMax = natToInteger @int == 0 || x1 > satPred SatBound maxBound
+    nF = natToNum @frac
+    yPlusHalf | nF == 0       = y
+              | isSigned yRep = y - (Fixed $ -1 `shiftL` (nF - 1))
+              | otherwise     = y + (Fixed $ 1 `shiftL` (nF - 1))
+
+  enumFromThenTo = enumFromThenTo#
+
+-- Inspired by Enum Int from GHC.Enum in base-4.14.1.0
+--
+-- Note that if x2 /= x1, it is guaranteed that (int + frac) >= 1, because if it
+-- were zero there would only be one concrete value. This fact is relied upon in
+-- enumFromThenToUp and enumFromThenToDown, which would have undefined behavior
+-- for (int + frac) == 0.
+enumFromThenTo#
+  :: forall f rep int frac
+   . ( NumFixedC rep int frac
+     , f ~ Fixed rep int frac)
+  => f
+  -> f
+  -> f
+  -> [f]
+enumFromThenTo# x1 x2 y
+  | x2 == x1  = if y < x1 then
+                  []
+                else
+                  repeat x1
+  | x2 > x1   = enumFromThenToUp x1 x2 y
+  | otherwise = enumFromThenToDown x1 x2 y
+
+enumFromThenToUp
+  :: forall f rep int frac
+   . ( NumFixedC rep int frac
+     , f ~ Fixed rep int frac)
+  => f
+  -> f
+  -> f
+  -> [f]
+enumFromThenToUp x1 x2 y
+  | y < x1 = let !y' = satAdd SatWrap y halfDelta  -- Never wraps
```

Comment:
```haskell
let !x = ....
```

basically means "when you evaluate `x` to WHNF, evaluate `x` to WHNF", which is superfluous. Bang patterns are useful in arguments, e.g.:


```haskell
foo !a !b = c
  where
    c = ....
```

where it would mean "when you evaluate the output of `foo`, first evaluate `a`, then `b`, and finally the output to WHNF. It's typically used to prevent huge chains of chunks from building up in accumulation parameters (which I think you mentioned during the call).

Reply:
I've dropped all bangs since we discovered the reason for them in `Int` is unrelated.